doc/man/man8/go500.8
.TH GO500 8C "20 August 2000" "OpenLDAP LDVERSION"
.\" $OpenLDAP$
.\" Copyright 1998-2002 The OpenLDAP Foundation All Rights Reserved.
.\" Copying restrictions apply.  See COPYRIGHT/LICENSE.
.SH NAME
go500 \- Local Gopher index search to X.500 search gateway
.SH SYNOPSIS
.B LIBEXECDIR/go500 [\-b searchbase] [\-d level] [\-l]
.B [\-x hostname] [\-c rdncount] [\-f filterfile]
.B [\-t templatefile]
.B [\-p port] [\-I]
.SH DESCRIPTION
.I go500
is the local gopher index search to X.500 search daemon.
It looks like a gopher index server to a gopher
client, translating the search criteria it is given into a
search of a pre-configured portion of the X.500 directory.
It uses LDAP to talk to X.500.  By default, it listens on
port 5555 for connections from gopher clients.
.LP
The
.B go500
server can be run either from
.BR inetd (8)
or as stand-alone server.
.SH STAND-ALONE OPERATION
To start
.B go500
as a stand-alone server, simply start it with no arguments
.LP
.nf
.ft tt
	LIBEXECDIR/go500
.ft
.fi
.LP
If you would like to start it at boot time add some lines like this to
the
.B etc/rc.local
or equivalent file:
.LP
.nf
.ft tt
	if [ -f LIBEXECDIR/go500 ]; then
		LIBEXECDIR/go500; echo ' go500'
	fi
.ft
.fi
.SH OPERATION WITH INETD
To arrange to have
.B go500
started from
.BR inetd (8),
the Internet protocol
daemon, add a line like the following to your
.B /etc/services
file, or the equivalent:
.LP
.nf
.ft tt
    go500           5555/tcp        go500
.ft
.fi
.LP
Next, add a line like this to your
.B /etc/inetd.conf
file, or the equivalent:
.LP
.nf
.ft tt
    go500   stream  tcp     nowait  nobody  LIBEXECDIR/go500    go500 -I
.ft
.fi
.LP
For these changes to take effect with inetd, you will probably have
to send it it a HUP signal.  See
.BR inetd (8)
for more details.
.SH GOPHER CONFIGURATION
The next step is to configure your local gopher server to have an
entry for
.BR go500 .
With the standard unix gopher server, this
can be done with a
.B .link
file.
A sample
.B .link
file is given below, with the things you should
change given in <>'s:
.LP
.nf
.ft tt
	Name=<Label of your choice>
	Type=7
	Port=5555
	Path=
	Host=<host.running.go500.here>
.ft
.fi
.LP
You may also have to restart your gopher daemon, or remove the
.B .cache
file.
See
.BR gopherd (8)
for more details.
.SH OPTIONS
.TP
.BI \-b " searchbase"
Specify an alternate starting point for searches.  The argument should
be a Distinguished Name in the form defined by RFC 1485.  For example,
the DN "o=University of Michigan, c=US" could be given to search the
University of Michigan portion of the X.500 tree.
.TP
.BI \-d " level"
Turn on debugging as defined by
.I level.
If this option is specified,
.B go500
will not fork or disassociate from the invoking terminal.  Some general
operation and status messages are printed for any value of \fIlevel\fP.
\fIlevel\fP is taken as a bit string, with each bit corresponding to a
different kind of debugging information.
.TP
.BI \-f " filterfile"
Specify an alternate filter configuration file for use with the
.BR ldap_getfilter (3)
facility, used by
.BR go500 .
.TP
.B \-l
Enable logging of various status and errors to the LOG_LOCAL3 facility via
syslog(8).
.TP
.BI \-p " port"
Specify an alternate port on which to listen for connections.
.TP
.BI \-t " templatefile"
Specify an alternate template  configuration  file  for
use  with  the
.BR ldap_init_templates (3)
facility, used by
.BR go500 .
.TP
.BI \-c " rdncount"
Specify the number of DN components to show for the names and DN attributes
within entries matching the search.
.TP
.BI \-x " hostname"
Specify an alternate host on which the ldap server is running.
.TP
.B \-I
Run from
.BR inetd (8).
.SH NOTES
Some implementations of inetd have a small limit on the number of arguments
that can be specified in the
.BR /etc/inetd.conf (5)
file.  This can cause
problems if you are using
.B go500
with a lot of arguments.
.LP
The default values for most of the things you can specify with
options are configured at compile time in the
.B include/ldapconfig.h.edit
include file.
.SH "SEE ALSO"
.BR ldap (3),
.BR inetd (8),
.BR gopherd (8),
.BR go500gw (8)
.SH ACKNOWLEDGEMENTS
.B	OpenLDAP
is developed and maintained by The OpenLDAP Project (http://www.openldap.org/).
.B	OpenLDAP
is derived from University of Michigan LDAP 3.3 Release.  


CHANGES
OpenLDAP 2.0 Change Log

OpenLDAP 2.0.X Engineering
	Fixed -lldap TLS data ready bug (ITS#821)
	Fixed -lldap abandon unsent op bug (ITS#837)
	Fixed -lldap URL search w/o host bug (ITS#843)
	Fixed -lldap referral handling bugs (ITS#799 ITS#817)
	Fixed slapd spasswd mutex bug
	Fixed slapd ACL nameuid bug
	Fixed slapd SASL layering bug
	Fixed slapd unknown authc method bug (ITS#831)
	Fixed slapd TLSVerifyClient config bug
	Updated -lldap TLS/SASL error reporting
	Updated -llber large element handling
	Updated slapadd error reporting
	Updated slapd chroot handling (ITS#810)
	Updated slapd subschema subentry
	Added slapd numericString indexing
	Removed lint
	Build Environment
		Fixed make comment bug (ITS#811)
		Fixed $(DESTDIR) install (ITS#806)
		Added proxy check to passwd test
		Fixed slurpd tempdir perms (ITS#840)
	Documentation
		Updated ldappasswd(1)
		Updated ldap_url(3)
		Updated slapd.conf(5)
		Updated slapd(8)
		Updated slurpd(8)

OpenLDAP 2.0.6 Release
	Build Environment
		Fixed schema install (ITS#802)
		Added libtool customizations
		Re-enabled module support
	Documentation
		Update ldapsearch(1)

OpenLDAP 2.0.5 Release
	Fixed slapd/ldbm/search confusing log message (ITS#771)
	Fixed slapd/sasl no cred bug
	Fixed slapd idletimeout lock bug (ITS#767)
	Fixed slapd caseIgnoreIA5string indexing bug (ITS#786)
	Fixed slurpd TLS initialization bug (ITS#768)
	Fixed slurpd empty one shot hang (ITS#798)
	Fixed -lldap SASL ctx close bug (ITS#790)
	Fixed -lldap first/next attribute bug (ITS#778)
	Fixed -lldap (v2) rebind URL bug
	Fixed -lldap msgfree NULL bug (ITS#800)
	Fixed -lldif empty value bug
	Fixed -llber stringal len=0 bug
	Updated -llber bprint routine
	Build Environment
		Fixed LDFLAGS/-rpath usage
		Fixed IPv6 detection (ITS#669,ITS#770)
		Fixed VPATH schema install
		Updated libtool to 1.3.5
		Added Darwin support
		Disabled module support

OpenLDAP 2.0.4 Release
	Fixed clients printf/usage bugs (ITS#749)
	Fixed -lldap SASL interoperability bug
	Fixed -lldap PF_LOCAL declaration/call bugs
	Fixed slapd SASL log error (ITS#759)
	Fixed slapd spasswd support (ITS#751)
	Fixed slapd/tools fixed sasl_props bug (ITS#761)
	Updated slurpd SASL support
	Added slapd accept(2) sockopt workaround (ITS#760)
	Added slurpd TLS support
	Removed lint
	Build Environment
		Fixed --enable-spasswd (ITS#751)
	Documentation
		Fixed ldif(5) file:/// bug (ITS#750)
		Updated ldap_schema(3)

OpenLDAP 2.0.3 Release
	Fixed -lldap KBIND typo (ITS#717)
	Fixed slapd ;binary search request subtyping

OpenLDAP 2.0.2 Release
	Fixed KBIND (ITS#717)
	Fixed clients/tools -R handling (ITS#726)
	Fixed -lldap SASL got_max_ssf bug 
	Fixed -lldap TCP_NODELAY bug 
	Fixed -lldap TLS issues (ITS#733)
	Fixed -lldap_r NT threads (ITS#732)
	Fixed -llutil sa_flags never set bug
	Fixed ldappasswd -A -S crash (ITS#714)
	Fixed ldappasswd user argument usage
	Fixed slapd disallow bind_anon (ITS#721)
	Fixed slapd IPv6 issues (ITS#716)
	Fixed slapd MIT KPASSWD Compatibility (ITS#715)
	Fixed slapd time syntax routines (ITS#713)
	Fixed slapd/ldbm DB_PRIVATE not set bug (ITS#725)
	Fixed slapd/ldbm nextid reporting bug
	Fixed slapd syntaxes/mr routines (ITS#739)
	Fixed slurpd -r buffer overflow (ITS#722)
	Updated slapd rootDSE inappropriate op handling
	Added slapd syntax/mr routines
	Added slapd allow/disallow options
	Added slapd defaultSearchBase and DN verify (ITS#723)
	Removed lint
	Build Environment
		Added test000-rootdse
		Added Corba & Java Schema
		Updated MSVC projects for BDB 3.1
		Fixed Kerberos detection (ITS#717)
		Remove incompatible contribWare
	Documentation
		Fixed ldappasswd(1) usage
		Fixed ldapmodify(1) (ITS#719)
		Updated release documents (ITS#720)

OpenLDAP 2.0.1 Release
	Fixed StartTLS & ldaps:// client SDK
	Fixed slapd session close deadlock (ITS#704)
	Fixed slapd SSF ACLs
	Fixed slapd "entry" ACLs
	Updated slapd MAXARGS to 200
	Updated slapd nisNetgroupTriple validation routine (ITS#705)
	Updated slapd bootParameter validation routine (ITS#706)
	Updated slapd nameUID syntax to reuse DN routines for uniqueMember
	Updated client tool usage (ITS#710)
	Build Environment
		Fixed MSC tcp_close redefine
	Documentation
		Updated manual pages references
		Updated release documents
		Updated back-sql documents
		Updated "to do" list

OpenLDAP 2.0.0 Release
	Initial release of OpenLDAP 2.0

---
$OpenLDAP$


clients/gopher/go500.c
/* $OpenLDAP$ */
/*
 * Copyright (c) 1990 Regents of the University of Michigan.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms are permitted
 * provided that this notice is preserved and that due credit is given
 * to the University of Michigan at Ann Arbor. The name of the University
 * may not be used to endorse or promote products derived from this
 * software without specific prior written permission. This software
 * is provided ``as is'' without express or implied warranty.
 */

#include "portable.h"

#include <stdio.h>

#include <ac/stdlib.h>

#include <ac/ctype.h>
#include <ac/signal.h>
#include <ac/socket.h>
#include <ac/string.h>
#include <ac/syslog.h>
#include <ac/time.h>
#include <ac/unistd.h>
#include <ac/wait.h>

#include <ac/param.h>
#include <ac/setproctitle.h>

#ifdef HAVE_SYS_RESOURCE_H
#include <sys/resource.h>
#endif

#include <ldap.h>
#include <disptmpl.h>

#include "ldap_defaults.h"

#define ldap_debug debug
#include "ldap_log.h"

#include "lutil.h"


int	debug;
int	dosyslog;
int ldap_syslog;
int ldap_syslog_level;
int	inetd;
int	dtblsize;

char	*ldaphost = NULL;
char	*base = NULL;
int	rdncount = GO500_RDNCOUNT;
char	*filterfile = FILTERFILE;
char	*templatefile = TEMPLATEFILE;

char	myhost[MAXHOSTNAMELEN];
int	myport;

static void usage	( char *name ) LDAP_GCCATTR((noreturn));
static int  set_socket	(int port);
static RETSIGTYPE wait4child(int sig);
static void do_queries	(int s)    LDAP_GCCATTR((noreturn));
static void do_error	(FILE *fp, char *s);
static void do_search	(LDAP *ld, FILE *fp, char *buf);
static void do_read	(LDAP *ld, FILE *fp, char *dn);

static void
usage( char *name )
{
	fprintf( stderr, "usage: %s [-d debuglevel] [-f filterfile] [-t templatefile]\r\n\t[-a] [-l] [-p port] [-x ldaphost] [-b searchbase] [-c rdncount]\r\n", name );
	exit( EXIT_FAILURE );
}

int
main( int argc, char **argv )
{
	int			s, ns, rc;
	int			port = -1;
	int			i, pid;
	char			*myname;
	fd_set			readfds;
	struct hostent		*hp;
	struct sockaddr_in	from;
	socklen_t         	fromlen;

#if defined( LDAP_PROCTITLE ) && !defined( HAVE_SETPROCTITLE )
	/* for setproctitle */
	Argv = argv;
 	Argc = argc;
#endif

	while ( (i = getopt( argc, argv, "b:d:f:lp:c:t:x:I" )) != EOF ) {
		switch( i ) {
		case 'b':	/* searchbase */
			base = strdup( optarg );
			break;

		case 'd':	/* debug level */
			debug |= atoi( optarg );
			break;

		case 'f':	/* ldap filter file */
			filterfile = strdup( optarg );
			break;

		case 'l':	/* log via LOG_LOCAL3 */
			dosyslog = 1;
			break;

		case 'p':	/* port to listen to */
			port = atoi( optarg );
			break;

		case 'c':	/* number of DN components to show */
			rdncount = atoi( optarg );
			break;

		case 't':	/* ldap template file */
			templatefile = strdup( optarg );
			break;

		case 'x':	/* ldap server hostname */
			ldaphost = strdup( optarg );
			break;

		case 'I':	/* run from inetd */
			inetd = 1;
			break;

		default:
			usage( argv[0] );
		}
	}

#ifdef GO500_HOSTNAME
	strcpy( myhost, GO500_HOSTNAME );
#else
	if ( myhost[0] == '\0' && gethostname( myhost, sizeof(myhost)-1 )
	    == -1 ) {
		perror( "gethostname" );
		exit( EXIT_FAILURE );
	}
	myhost[sizeof(myhost)-1] = '\0';
#endif

#ifdef HAVE_SYSCONF
	dtblsize = sysconf( _SC_OPEN_MAX );
#elif HAVE_GETDTABLESIZE
	dtblsize = getdtablesize();
#else
	dtblsize = FD_SETSIZE;
#endif

#ifdef FD_SETSIZE
	if (dtblsize > FD_SETSIZE) {
		dtblsize = FD_SETSIZE;
	}
#endif	/* FD_SETSIZE*/

	/* detach if stderr is redirected or no debugging */
	if ( inetd == 0 )
		lutil_detach( debug && !isatty( 1 ), 1 );

	if ( (myname = strrchr( argv[0], '/' )) == NULL )
		myname = strdup( argv[0] );
	else
		myname = strdup( myname + 1 );

	if ( debug ) {
		ber_set_option(NULL, LDAP_OPT_DEBUG_LEVEL, &debug);
		ldap_set_option(NULL, LDAP_OPT_DEBUG_LEVEL, &debug);
	}

#ifdef SIGPIPE
	(void) SIGNAL( SIGPIPE, SIG_IGN );
#endif

	if ( dosyslog ) {
#ifdef LOG_LOCAL3
		openlog( myname, OPENLOG_OPTIONS, LOG_LOCAL3 );
#elif LOG_DEBUG
		openlog( myname, OPENLOG_OPTIONS );
#endif
	}
	if ( dosyslog )
		syslog( LOG_INFO, "initializing" );

	/* set up the socket to listen on */
	if ( inetd == 0 ) {
		s = set_socket( port );

		/* arrange to reap children */
		(void) SIGNAL( SIGCHLD, wait4child );
	} else {
		myport = GO500_PORT;

		fromlen = sizeof(from);
		if ( getpeername( 0, (struct sockaddr *) &from, &fromlen )
		    == 0 ) {
			hp = gethostbyaddr( (char *) &(from.sin_addr),
			    sizeof(from.sin_addr), AF_INET );
			Debug( LDAP_DEBUG_ARGS, "connection from %s (%s)\n",
			    (hp == NULL) ? "unknown" : hp->h_name,
			    inet_ntoa( from.sin_addr ), 0 );

			if ( dosyslog ) {
				syslog( LOG_INFO, "connection from %s (%s)",
				    (hp == NULL) ? "unknown" : hp->h_name,
				    inet_ntoa( from.sin_addr ) );
			}

#ifdef LDAP_PROCTITLE
			setproctitle( "%s", hp == NULL ? inet_ntoa( from.sin_addr ) :
			    hp->h_name );
#endif
		}

		do_queries( 0 );

		exit( EXIT_SUCCESS );
	}

	for ( ;; ) {
		FD_ZERO( &readfds );
		FD_SET( s, &readfds );

		if ( (rc = select( dtblsize, &readfds, 0, 0 ,0 )) == -1 ) {
			if ( debug ) perror( "select" );
			continue;
		} else if ( rc == 0 ) {
			continue;
		}

		if ( ! FD_ISSET( s, &readfds ) )
			continue;

		fromlen = sizeof(from);
		if ( (ns = accept( s, (struct sockaddr *) &from, &fromlen ))
		    == -1 ) {
			if ( debug ) perror( "accept" );
			exit( EXIT_FAILURE );
		}

		hp = gethostbyaddr( (char *) &(from.sin_addr),
		    sizeof(from.sin_addr), AF_INET );

		if ( dosyslog ) {
			syslog( LOG_INFO, "TCP connection from %s (%s)",
			    (hp == NULL) ? "unknown" : hp->h_name,
			    inet_ntoa( from.sin_addr ) );
		}

		switch( pid = fork() ) {
		case 0:		/* child */
			tcp_close( s );
			do_queries( ns );
			break;

		case -1:	/* failed */
			perror( "fork" );
			break;

		default:	/* parent */
			tcp_close( ns );
			if ( debug )
				fprintf( stderr, "forked child %d\n", pid );
			break;
		}
	}
	/* NOT REACHED */
}

static int
set_socket( int port )
{
	int			s, one;
	struct sockaddr_in	addr;

	if ( port == -1 )
		port = GO500_PORT;
	myport = port;

	if ( (s = socket( AF_INET, SOCK_STREAM, 0 )) == -1 ) {
                perror( "socket" );
                exit( EXIT_FAILURE );
        }

#ifdef SO_REUSEADDR
        /* set option so clients can't keep us from coming back up */
		one = 1;
        if ( setsockopt( s, SOL_SOCKET, SO_REUSEADDR, (char *) &one,
	    sizeof(one) ) < 0 ) {
                perror( "setsockopt" );
                exit( EXIT_FAILURE );
        }
#endif
#ifdef SO_KEEPALIVE
		/* enable keep alives */
		one = 1;
        if ( setsockopt( s, SOL_SOCKET, SO_KEEPALIVE, (char *) &one,
	    sizeof(one) ) < 0 ) {
                perror( "setsockopt" );
                exit( EXIT_FAILURE );
        }
#endif

        /* bind to a name */
        addr.sin_family = AF_INET;
        addr.sin_addr.s_addr = htonl(INADDR_ANY);
        addr.sin_port = htons( port );
        if ( bind( s, (struct sockaddr *) &addr, sizeof(addr) ) ) {
                perror( "bind" );
                exit( EXIT_FAILURE );
        }

	/* listen for connections */
        if ( listen( s, 5 ) == -1 ) {
                perror( "listen" );
                exit( EXIT_FAILURE );
        }

        if ( debug ) printf("tcp socket allocated, bound, and listening\n");

	return( s );
}

static RETSIGTYPE
wait4child( int sig )
{
#ifndef HAVE_WAITPID
        WAITSTATUSTYPE     status;
#endif

        if ( debug ) printf( "parent: catching child status\n" );

#ifdef HAVE_WAITPID
	while (waitpid ((pid_t) -1, (int *) NULL, WAIT_FLAGS) > 0)
		;	/* NULL */
#else
	while ( wait4((pid_t) -1, &status, WAIT_FLAGS, 0 ) > 0 )
		;	/* NULL */
#endif

	(void) SIGNAL_REINSTALL ( SIGCHLD, wait4child );
}

static void
do_queries( int s )
{
	char		buf[1024], *query;
	int		len;
	FILE		*fp;
	int		rc;
	struct timeval	timeout;
	fd_set		readfds;
	LDAP		*ld;

	if ( (fp = fdopen( s, "a+")) == NULL ) {
		exit( EXIT_FAILURE );
	}

	timeout.tv_sec = GO500_TIMEOUT;
	timeout.tv_usec = 0;
	FD_ZERO( &readfds );
	FD_SET( fileno( fp ), &readfds );

	if ( (rc = select( dtblsize, &readfds, 0, 0, &timeout )) <= 0 )
		exit( EXIT_FAILURE );

	if ( fgets( buf, sizeof(buf), fp ) == NULL )
		exit( EXIT_FAILURE );

	len = strlen( buf );
	if ( debug ) {
		fprintf( stderr, "got %d bytes\n", len );
#ifdef LDAP_DEBUG
		ber_bprint( buf, len );
#endif
	}

	/* strip of \r \n */
	if ( buf[len - 1] == '\n' )
		buf[len - 1] = '\0';
	len--;
	if ( buf[len - 1] == '\r' )
		buf[len - 1] = '\0';
	len--;

	query = buf;

	/* strip off leading white space */
	while ( isspace( (unsigned char) *query )) {
		++query;
		--len;
	}

	rewind(fp);

	if ( *query == '~' || *query == '@' ) {
		ld = NULL;
	} else if ( (ld = ldap_init( ldaphost, 0 )) == NULL ) {
		fprintf(fp,
			"0An error occurred (explanation)\t@%d\t%s\t%d\r\n",
			LDAP_SERVER_DOWN, myhost, myport );
		fprintf( fp, ".\r\n" );
		rewind(fp);
		exit( EXIT_FAILURE );
	} else {
		int deref = GO500_DEREF;
		ldap_set_option(ld, LDAP_OPT_DEREF, &deref);

		rc = ldap_simple_bind_s( ld, NULL, NULL );
		if ( rc != LDAP_SUCCESS ) {
			fprintf(fp,
			    "0An error occurred (explanation)\t@%d\t%s\t%d\r\n",
			    rc, myhost, myport );
			fprintf( fp, ".\r\n" );
			rewind(fp);
			exit( EXIT_FAILURE );
		}
	}

	switch ( *query ) {
	case '~':
		fprintf( fp, "The query you specified was not specific enough, causing a size limit\r\n" );
		fprintf( fp, "to be exceeded and the first several matches found to be returned.\r\n" );
		fprintf( fp, "If you did not find the match you were looking for, try issuing a more\r\n" );
		fprintf( fp, "specific query, for example one that contains both first and last name.\r\n" );
		fprintf( fp, ".\r\n" );
		break;

	case '=':
		do_read( ld, fp, ++query );
		break;

	case '@':
		do_error( fp, ++query );
		break;

	default:
		do_search( ld, fp, query );
		break;
	}

	fprintf( fp, ".\r\n" );
	rewind(fp);

	if ( ld != NULL) {
		ldap_unbind( ld );
	}

	exit( EXIT_FAILURE );
	/* NOT REACHED */
}

static void
do_error( FILE *fp, char *s )
{
	int	code;

	code = atoi( s );

	fprintf( fp, "An error occurred searching X.500.  The error code was %d\r\n", code );
	fprintf( fp, "The corresponding error is: %s\r\n", ldap_err2string( code ) );
	fprintf( fp, "No additional information is available\r\n" );
	fprintf( fp, ".\r\n" );
}

static void
do_search( LDAP *ld, FILE *fp, char *buf )
{
	char		*dn, *rdn;
	char		**title;
	int		rc, matches = 0;
	struct timeval	tv;
	LDAPFiltInfo	*fi;
	LDAPFiltDesc	*filtd;
	LDAPMessage	*e, *res;
	static char	*attrs[] = { "title", 0 };

		if ( (filtd = ldap_init_getfilter( filterfile )) == NULL ) {
			fprintf( stderr, "Cannot open filter file (%s)\n",
			    filterfile );
			exit( EXIT_FAILURE );
		}

		tv.tv_sec = GO500_TIMEOUT;
		tv.tv_usec = 0;
		for ( fi = ldap_getfirstfilter( filtd, "go500", buf );
		    fi != NULL;
		    fi = ldap_getnextfilter( filtd ) )
		{
			if ( (rc = ldap_search_st( ld, base, LDAP_SCOPE_SUBTREE,
			    fi->lfi_filter, attrs, 0, &tv, &res ))
			    != LDAP_SUCCESS && rc != LDAP_SIZELIMIT_EXCEEDED ) {
				fprintf(fp, "0An error occurred (explanation)\t@%d\t%s\t%d\r\n",
				    rc, myhost, myport );
				ldap_getfilter_free( filtd );
				return;
			}

			if ( (matches = ldap_count_entries( ld, res )) != 0 )
				break;
		}
		ldap_getfilter_free( filtd );

	if ( matches <= 0 ) {
		return;
	}

#ifdef GO500_SORT_ATTR
	ldap_sort_entries( ld, &res, GO500_SORT_ATTR, strcasecmp );
#endif

	for ( e = ldap_first_entry( ld, res ); e != NULL;
	    e = ldap_next_entry( ld, e ) ) {
		char	*s;

		dn = ldap_get_dn( ld, e );
		rdn = strdup( dn );
		if ( (s = strchr( rdn, ',' )) != NULL )
			*s = '\0';

		if ( (s = strchr( rdn, '=' )) == NULL )
			s = rdn;
		else
			++s;

		title = ldap_get_values( ld, e, "title" );

		if ( title != NULL ) {
			char	*p;

			for ( p = title[0]; *p; p++ ) {
				if ( *p == '/' )
					*p = '\\';
			}
		}

		fprintf( fp, "0%-20s    %s\t=%s\t%s\t%d\r\n", s,
		    title ? title[0] : "", dn, myhost, myport );

		if ( title != NULL )
			ldap_value_free( title );

		free( rdn );
		free( dn );
	}

	if ( ldap_result2error( ld, res, 1 ) == LDAP_SIZELIMIT_EXCEEDED ) {
		fprintf( fp, "0A size limit was exceeded (explanation)\t~\t%s\t%d\r\n",
		    myhost, myport );
	}
}

static int
entry2textwrite( void *fp, char *buf, ber_len_t len )
{
	return( fwrite( buf, len, 1, (FILE *)fp ) == 0 ? -1 : len );
}

static void
do_read( LDAP *ld, FILE *fp, char *dn )
{
	static struct ldap_disptmpl *tmpllist;

	ldap_init_templates( templatefile, &tmpllist );

	if ( ldap_entry2text_search( ld, dn, base, NULL, tmpllist, NULL, NULL,
	    entry2textwrite, (void *) fp, "\r\n", rdncount,
	    LDAP_DISP_OPT_DOSEARCHACTIONS ) != LDAP_SUCCESS ) {
		ldap_perror( ld, "ldap_entry2text_search" );
		exit( EXIT_FAILURE );
	}

	if ( tmpllist != NULL ) {
		ldap_free_templates( tmpllist );
	}
}
